README.md
# pg-dba
[![CI](https://github.com/jonstacks/pg-dba/actions/workflows/ci.yml/badge.svg)](https://github.com/jonstacks/pg-dba/actions/workflows/ci.yml)
[![codecov](https://codecov.io/gh/jonstacks/pg-dba/branch/master/graph/badge.svg)](https://codecov.io/gh/jonstacks/pg-dba)

A semi-automated Postgres DBA for helping teams better manage their DB.

## Testing

To run integration tests locally, you must have docker installed. You can run
the integration tests by running:

```
make compose-up
make integration
```

## Docs

You can run the docs server by going to the `docs` folder and running:

```
make doc-server
```

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.3.0] - 2021-07-03

### Added

- Environment variable to configure bloat query timeout
- CHANGELOG.md

### Changed

- Moved to GitHub Actions for CI.
- Postgres version in docker-compose.yaml
- Golang to 1.16
- Updated documentation


pkg/config/config_test.go
package config

import (
	"os"
	"testing"

	"github.com/sirupsen/logrus"

	"github.com/stretchr/testify/assert"
)

func withEnvSet(key, value string, f func(key string)) {
	os.Setenv(key, value)
	f(key)
	os.Unsetenv(key)
}

func testAllBoolVariable(t *testing.T, key string, defaultValue bool, f func() bool) {
	for _, v := range []string{"true", "TRUE", "TrUE"} {
		withEnvSet(key, v, func(k string) {
			assert.True(t, f())
		})
	}

	for _, v := range []string{"false", "FALSE", "FaLSE"} {
		withEnvSet(key, v, func(k string) {
			assert.False(t, f())
		})
	}

	assert.Equal(t, defaultValue, f())
}

func testAllStringVariable(t *testing.T, key string, defaultValue string, f func() string) {
	assert.Equal(t, defaultValue, f())

	for _, v := range []string{"value one", "value_two", "VALUE_THREE"} {
		withEnvSet(key, v, func(k string) {
			assert.Equal(t, v, f())
		})
	}
}

func testAllTimeoutVariable(t *testing.T, key string, defaultValue int, f func() (int, error)) {
	val, err := f()
	assert.NoError(t, err)
	assert.Equal(t, defaultValue, val)

	for intVal, stringVal := range map[int]string{100: "100", 200: "200", 1000: "1000"} {
		withEnvSet(key, stringVal, func(k string) {
			val, err := f()
			assert.NoError(t, err)
			assert.Equal(t, intVal, val)
		})
	}

	withEnvSet(key, "SHOULD NOT PARSE", func(k string) {
		_, err := f()
		assert.Error(t, err)
	})
}

func TestEnvBool(t *testing.T) {
	withEnvSet("BOOLEAN_KEY", "wrong", func(key string) {
		val := envBool(key, true)
		assert.False(t, val)

		val = envBool(key, false)
		assert.False(t, val)
	})

	withEnvSet("BOOLEAN_KEY", "TRUE", func(key string) {
		val := envBool(key, false)
		assert.True(t, val)

		val = envBool(key, true)
		assert.True(t, val)
	})

	// No Value set
	val := envBool("BOOLEAN_KEY", false)
	assert.False(t, val)

	val = envBool("BOOLEAN_KEY", true)
	assert.True(t, val)
}

func TestEnvInt(t *testing.T) {
	withEnvSet("MYINT", "10000", func(key string) {
		val, err := envInt(key, 3000)
		assert.NoError(t, err)
		assert.Equal(t, 10000, val)
	})

	withEnvSet("MYINT", "", func(key string) {
		val, err := envInt(key, 1000)
		assert.NoError(t, err)
		assert.Equal(t, 1000, val)
	})
}

func TestEnvDefault(t *testing.T) {
	withEnvSet("TEST_VAR", "SomeValue", func(key string) {
		assert.Equal(t, "SomeValue", envDefault(key, "Default Value"))
	})

	assert.Equal(t, "Default Value", envDefault("TEST_VAR", "Default Value"))
}

func TestLogFormat(t *testing.T) {
	withEnvSet("LOG_FORMAT", "json", func(key string) {
		assert.IsType(t, &logrus.JSONFormatter{}, LogFormat())
	})

	withEnvSet("LOG_FORMAT", "text", func(key string) {
		assert.IsType(t, &logrus.TextFormatter{}, LogFormat())
	})

	assert.IsType(t, &logrus.TextFormatter{}, LogFormat())
}

func TestLogLevel(t *testing.T) {
	withEnvSet("LOG_LEVEL", "DEBUG", func(key string) {
		assert.Equal(t, logrus.DebugLevel, LogLevel())
	})

	withEnvSet("LOG_LEVEL", "WARN", func(key string) {
		assert.Equal(t, logrus.WarnLevel, LogLevel())
	})

	withEnvSet("LOG_LEVEL", "Error", func(key string) {
		assert.Equal(t, logrus.ErrorLevel, LogLevel())
	})

	withEnvSet("LOG_LEVEL", "INFO", func(key string) {
		assert.Equal(t, logrus.InfoLevel, LogLevel())
	})

	assert.Equal(t, logrus.InfoLevel, LogLevel())
}

func TestAnalyzeTimeoutSeconds(t *testing.T) {
	testAllTimeoutVariable(t, "ANALYZE_TIMEOUT_SECONDS", 600, AnalyzeTimeoutSeconds)
}

func TestFullVacuumTimeoutSeconds(t *testing.T) {
	testAllTimeoutVariable(t, "FULL_VACUUM_TIMEOUT_SECONDS", 600, FullVacuumTimeoutSeconds)
}

func TestVacuumTimeoutSeconds(t *testing.T) {
	testAllTimeoutVariable(t, "VACUUM_TIMEOUT_SECONDS", 600, VacuumTimeoutSeconds)
}

func TestBloatQueryTimeoutSeconds(t *testing.T) {
	testAllTimeoutVariable(t, "BLOAT_QUERY_TIMEOUT_SECONDS", 30, BloatQueryTimeoutSeconds)
}

func TestPostAnalyze(t *testing.T) {
	testAllBoolVariable(t, "POST_ANALYZE", true, PostAnalyze)
}

func TestPreAnalyze(t *testing.T) {
	testAllBoolVariable(t, "PRE_ANALYZE", true, PreAnalyze)
}

func TestVerbose(t *testing.T) {
	testAllBoolVariable(t, "VERBOSE", false, Verbose)
}

func TestDBHost(t *testing.T) {
	testAllStringVariable(t, "POSTGRES_HOST", "localhost", DBHost)
}

func TestDBUser(t *testing.T) {
	testAllStringVariable(t, "POSTGRES_USER", "postgres", DBUser)
}

func TestDBName(t *testing.T) {
	testAllStringVariable(t, "POSTGRES_DB", "postgres", DBName)
}

func TestDBPassword(t *testing.T) {
	testAllStringVariable(t, "POSTGRES_PASSWORD", `""`, DBPassword)
}

func TestDBConnectionString(t *testing.T) {
	withEnvSet("SSL_MODE", "require", func(key string) {
		assert.Equal(t, `host=localhost dbname=postgres user=postgres password="" sslmode=require`, DBConnectionString())
	})
}


pkg/dba/dba_integration_test.go
// +build integration

package dba

import (
	"testing"
	"time"

	"github.com/jonstacks/pg-dba/pkg/config"
	"github.com/stretchr/testify/assert"
)

// Tests The automatic DBA
func TestDBACase1(t *testing.T) {
	opts := NewDefaultOptions()
	opts.Verbose = config.Verbose()
	dba := New(config.DBConnectionString(), opts)
	assert.Nil(t, dba.Run())
}

func TestDBANonVerbose(t *testing.T) {
	opts := NewDefaultOptions()
	opts.Verbose = false
	dba := New(config.DBConnectionString(), opts)
	assert.Nil(t, dba.Run())
}

func TestDBAAnalyzeTimeout(t *testing.T) {
	opts := NewDefaultOptions()
	opts.AnalyzeTimeout = 50 * time.Millisecond
	opts.Verbose = config.Verbose()
	dba := New(config.DBConnectionString(), opts)
	assert.Error(t, dba.Run())
}


pkg/dba/dba.go
package dba

import (
	"context"
	"database/sql"
	"fmt"
	"time"

	"github.com/jonstacks/pg-dba/pkg/utils"
	_ "github.com/lib/pq" // Import postgres db driver
	"github.com/sirupsen/logrus"
)

// Options controll how the DBA determines what to do
type Options struct {
	PreAnalyze     bool
	AnalyzeTimeout time.Duration
	PostAnalyze    bool

	BloatQueryTimeout time.Duration

	PreVacuum    bool
	VacuumTimout time.Duration

	FullVacuumBloatPercent   int           // Only consider tables with bloat larger than this percent to need a full vacuum
	FullVacuumMaxTableSizeMb int           // Do not vacuum tables larger than this size due to the amount of time it would take
	FullVacuumTimeout        time.Duration // Cancel full vacuum's that take longer than this many seconds

	Timeout time.Duration // Default timeout for everything else
	Verbose bool
}

// NewDefaultOptions returns new Options initilized with our reasonable defaults
func NewDefaultOptions() Options {
	return Options{
		PreAnalyze:     true,
		AnalyzeTimeout: 10 * time.Minute,
		PostAnalyze:    true,

		BloatQueryTimeout: 30 * time.Second,

		PreVacuum:    false,
		VacuumTimout: 10 * time.Minute,

		FullVacuumBloatPercent:   10,
		FullVacuumMaxTableSizeMb: 10 * 1000, // Only Vacuum tables smaller than 10 GB by default
		FullVacuumTimeout:        10 * time.Minute,

		Timeout: 30 * time.Second,
		Verbose: false,
	}
}

// DBA is an automatic database administrator.
type DBA struct {
	connStr string
	db      *sql.DB
	options Options
}

// New creates a new DBA and returns it
func New(connStr string, opts Options) *DBA {
	return &DBA{
		connStr: connStr,
		options: opts,
	}
}

// Run runs the automatic DBA
func (dba *DBA) Run() error {
	var err error

	logrus.Info("Connecting to database")
	dba.db, err = sql.Open("postgres", dba.connStr)
	if err != nil {
		return err
	}
	defer dba.db.Close()

	if err = dba.db.Ping(); err != nil {
		return err
	}

	logrus.Info("Running Auto DBA")

	if dba.options.PreVacuum {
		logrus.Info("Running a vacuum to clean up space")
		if err = dba.vacuum(dba.options.PreAnalyze); err != nil {
			return err
		}
	} else if dba.options.PreAnalyze {
		logrus.Info("Running pre-analyze")
		if err = dba.analyze(); err != nil {
			return err
		}
	}

	// Now let's see which tables need a full vacuum
	results, err := dba.bloatedTables()
	if err != nil {
		return err
	}

	for _, r := range results {
		logContext := logrus.WithFields(logrus.Fields{
			"table_mb":      r.TableMb,
			"table_name":    r.TableName,
			"schema_name":   r.SchemaName,
			"percent_bloat": r.PercentBloat,
		})
		if r.TableMb <= float64(dba.options.FullVacuumMaxTableSizeMb) {
			logContext.Info("Table can be vacuumed.")
			err = dba.fullVacuum(fmt.Sprintf("%s.%s", r.SchemaName, r.TableName))
			if err != nil {
				return err
			}
		} else {
			logContext.Warn("Table can NOT be vacuumed due to table size.")
		}
	}

	if dba.options.PostAnalyze {
		logrus.Info("Running post-analyze to update statistics for query planner")
		err = dba.analyze()
	}

	return err
}

func (dba *DBA) analyze() error {
	var cmd string
	if dba.options.Verbose {
		cmd = "ANALYZE VERBOSE"
	} else {
		cmd = "ANALYZE"
	}

	ctx, cancel := context.WithTimeout(context.Background(), dba.options.AnalyzeTimeout)
	defer cancel()

	_, err := dba.execContext(ctx, cmd)
	return err
}

func (dba *DBA) fullVacuum(tableName string) error {
	cmd := fmt.Sprintf("VACUUM (FULL, ANALYZE) %s", tableName)
	ctx, cancel := context.WithTimeout(context.Background(), dba.options.VacuumTimout)
	defer cancel()
	_, err := dba.execContext(ctx, cmd)
	return err
}

func (dba *DBA) vacuum(analyze bool) error {
	var cmd string
	if dba.options.Verbose {
		cmd = "VACUUM (VERBOSE, ANALYZE)"
	} else {
		cmd = "VACUUM ANALYZE"
	}

	ctx, cancel := context.WithTimeout(context.Background(), dba.options.VacuumTimout)
	defer cancel()
	_, err := dba.execContext(ctx, cmd)
	return err
}

func (dba *DBA) bloatedTables() ([]TableBloatResult, error) {
	logrus.Info("Running table bloat query")
	results := make([]TableBloatResult, 0)
	query := TableBloatQuery(dba.options.FullVacuumBloatPercent)
	ctx, cancel := context.WithTimeout(context.Background(), dba.options.BloatQueryTimeout)
	defer cancel()
	rows, err := dba.queryContext(ctx, query)
	if err != nil {
		return results, err
	}
	for rows.Next() {
		var r TableBloatResult
		if err := rows.Scan(&r.DatabaseName, &r.SchemaName, &r.TableName, &r.CanEstimate, &r.EstimatedRows, &r.PercentBloat, &r.MbBloat, &r.TableMb); err != nil {
			return results, err
		}
		results = append(results, r)
	}
	return results, nil
}

func (dba *DBA) queryContext(ctx context.Context, query string, args ...interface{}) (*sql.Rows, error) {
	var err error
	var rows *sql.Rows

	logrus.Debugf("Running '%s'", query)
	runTime := utils.Time(func() { rows, err = dba.db.QueryContext(ctx, query, args...) })
	logContext := logrus.WithFields(logrus.Fields{"duration": runTime.String()})
	if err != nil {
		logContext.Errorf("'%s' finished with error '%s'", query, err)
	} else {
		logContext.Infof("'%s' finished successfully", query)
	}
	return rows, err
}

func (dba *DBA) execContext(ctx context.Context, query string, args ...interface{}) (sql.Result, error) {
	var err error
	var result sql.Result

	logrus.Debugf("Running '%s'", query)
	runTime := utils.Time(func() { result, err = dba.db.ExecContext(ctx, query, args...) })
	logContext := logrus.WithFields(logrus.Fields{"duration": runTime.String()})
	if err != nil {
		logContext.Errorf("'%s' finished with error '%s'", query, err)
	} else {
		logContext.Infof("'%s' finished successfully", query)
	}
	return result, err
}


cmd/pg-dba/main.go
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/jonstacks/pg-dba/pkg/config"
	"github.com/jonstacks/pg-dba/pkg/dba"
	"github.com/sirupsen/logrus"
)

func fatal(err error) {
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
		os.Exit(1)
	}
}

func init() {
	logrus.SetLevel(config.LogLevel())
	logrus.SetFormatter(config.LogFormat())
}

func main() {

	analyzeTimeoutSecs, err := config.AnalyzeTimeoutSeconds()
	fatal(err)

	vacuumTimeoutSecs, err := config.VacuumTimeoutSeconds()
	fatal(err)

	fullVacuumTimeoutSecs, err := config.FullVacuumTimeoutSeconds()
	fatal(err)

	bloatQueryTimeoutSecs, err := config.BloatQueryTimeoutSeconds()
	fatal(err)

	opts := dba.NewDefaultOptions()
	opts.AnalyzeTimeout = time.Duration(analyzeTimeoutSecs) * time.Second
	opts.VacuumTimout = time.Duration(vacuumTimeoutSecs) * time.Second
	opts.FullVacuumTimeout = time.Duration(fullVacuumTimeoutSecs) * time.Second
	opts.BloatQueryTimeout = time.Duration(bloatQueryTimeoutSecs) * time.Second
	opts.Verbose = config.Verbose()
	admin := dba.New(config.DBConnectionString(), opts)
	fatal(admin.Run())
}


pkg/dba/queries.go
package dba

import "fmt"

// TableBloatResult returns the results of checking table bloat
type TableBloatResult struct {
	DatabaseName  string
	SchemaName    string
	TableName     string
	CanEstimate   bool
	EstimatedRows int
	PercentBloat  int
	MbBloat       float64
	TableMb       float64
}

// TableBloatQuery returns the query to run to check for table bloat. Taken
// from: https://raw.githubusercontent.com/pgexperts/pgx_scripts/master/bloat/table_bloat_check.sql
func TableBloatQuery(minPercentBloat int) string {
	return fmt.Sprintf(`-- new table bloat query
-- still needs work; is often off by +/- 20%%
WITH constants AS (
    -- define some constants for sizes of things
    -- for reference down the query and easy maintenance
    SELECT current_setting('block_size')::numeric AS bs, 23 AS hdr, 8 AS ma
),
no_stats AS (
    -- screen out table who have attributes
    -- which dont have stats, such as JSON
    SELECT table_schema, table_name,
        n_live_tup::numeric as est_rows,
        pg_table_size(relid)::numeric as table_size
    FROM information_schema.columns
        JOIN pg_stat_user_tables as psut
           ON table_schema = psut.schemaname
           AND table_name = psut.relname
        LEFT OUTER JOIN pg_stats
        ON table_schema = pg_stats.schemaname
            AND table_name = pg_stats.tablename
            AND column_name = attname
    WHERE attname IS NULL
        AND table_schema NOT IN ('pg_catalog', 'information_schema')
    GROUP BY table_schema, table_name, relid, n_live_tup
),
null_headers AS (
    -- calculate null header sizes
    -- omitting tables which dont have complete stats
    -- and attributes which aren't visible
    SELECT
        hdr+1+(sum(case when null_frac <> 0 THEN 1 else 0 END)/8) as nullhdr,
        SUM((1-null_frac)*avg_width) as datawidth,
        MAX(null_frac) as maxfracsum,
        schemaname,
        tablename,
        hdr, ma, bs
    FROM pg_stats CROSS JOIN constants
        LEFT OUTER JOIN no_stats
            ON schemaname = no_stats.table_schema
            AND tablename = no_stats.table_name
    WHERE schemaname NOT IN ('pg_catalog', 'information_schema')
        AND no_stats.table_name IS NULL
        AND EXISTS ( SELECT 1
            FROM information_schema.columns
                WHERE schemaname = columns.table_schema
                    AND tablename = columns.table_name )
    GROUP BY schemaname, tablename, hdr, ma, bs
),
data_headers AS (
    -- estimate header and row size
    SELECT
        ma, bs, hdr, schemaname, tablename,
        (datawidth+(hdr+ma-(case when hdr%%ma=0 THEN ma ELSE hdr%%ma END)))::numeric AS datahdr,
        (maxfracsum*(nullhdr+ma-(case when nullhdr%%ma=0 THEN ma ELSE nullhdr%%ma END))) AS nullhdr2
    FROM null_headers
),
table_estimates AS (
    -- make estimates of how large the table should be
    -- based on row and page size
    SELECT schemaname, tablename, bs,
        reltuples::numeric as est_rows, relpages * bs as table_bytes,
    CEIL((reltuples*
            (datahdr + nullhdr2 + 4 + ma -
                (CASE WHEN datahdr%%ma=0
                    THEN ma ELSE datahdr%%ma END)
                )/(bs-20))) * bs AS expected_bytes,
        reltoastrelid
    FROM data_headers
        JOIN pg_class ON tablename = relname
        JOIN pg_namespace ON relnamespace = pg_namespace.oid
            AND schemaname = nspname
    WHERE pg_class.relkind = 'r'
),
estimates_with_toast AS (
    -- add in estimated TOAST table sizes
    -- estimate based on 4 toast tuples per page because we dont have
    -- anything better.  also append the no_data tables
    SELECT schemaname, tablename,
        TRUE as can_estimate,
        est_rows,
        table_bytes + ( coalesce(toast.relpages, 0) * bs ) as table_bytes,
        expected_bytes + ( ceil( coalesce(toast.reltuples, 0) / 4 ) * bs ) as expected_bytes
    FROM table_estimates LEFT OUTER JOIN pg_class as toast
        ON table_estimates.reltoastrelid = toast.oid
            AND toast.relkind = 't'
),
table_estimates_plus AS (
-- add some extra metadata to the table data
-- and calculations to be reused
-- including whether we cant estimate it
-- or whether we think it might be compressed
    SELECT current_database() as databasename,
            schemaname, tablename, can_estimate,
            est_rows,
            CASE WHEN table_bytes > 0
                THEN table_bytes::NUMERIC
                ELSE NULL::NUMERIC END
                AS table_bytes,
            CASE WHEN expected_bytes > 0
                THEN expected_bytes::NUMERIC
                ELSE NULL::NUMERIC END
                    AS expected_bytes,
            CASE WHEN expected_bytes > 0 AND table_bytes > 0
                AND expected_bytes <= table_bytes
                THEN (table_bytes - expected_bytes)::NUMERIC
                ELSE 0::NUMERIC END AS bloat_bytes
    FROM estimates_with_toast
    UNION ALL
    SELECT current_database() as databasename,
        table_schema, table_name, FALSE,
        est_rows, table_size,
        NULL::NUMERIC, NULL::NUMERIC
    FROM no_stats
),
bloat_data AS (
    -- do final math calculations and formatting
    select current_database() as databasename,
        schemaname, tablename, can_estimate,
        table_bytes, round(table_bytes/(1024^2)::NUMERIC,3) as table_mb,
        expected_bytes, round(expected_bytes/(1024^2)::NUMERIC,3) as expected_mb,
        round(bloat_bytes*100/table_bytes) as pct_bloat,
        round(bloat_bytes/(1024::NUMERIC^2),2) as mb_bloat,
        table_bytes, expected_bytes, est_rows
    FROM table_estimates_plus
)
-- filter output for bloated tables
SELECT databasename, schemaname, tablename,
    can_estimate,
    est_rows,
    pct_bloat, mb_bloat,
    table_mb
FROM bloat_data
-- this where clause defines which tables actually appear
-- in the bloat chart
-- example below filters for tables which are either 50%%
-- bloated and more than 20mb in size, or more than 25%%
-- bloated and more than 4GB in size
WHERE pct_bloat >= %d
ORDER BY pct_bloat DESC;`, minPercentBloat)
}


pkg/config/config.go
package config

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/sirupsen/logrus"
)

func envBool(name string, defaultValue bool) bool {
	var defaultString string

	if defaultValue {
		defaultString = "true"
	} else {
		defaultString = "false"
	}
	return lowerEnvDefault(name, defaultString) == "true"
}

func envDefault(name, defaultValue string) string {
	val := os.Getenv(name)
	if val == "" {
		return defaultValue
	}
	return val
}

func lowerEnvDefault(name, defaultValue string) string {
	return strings.ToLower(envDefault(name, defaultValue))
}

func envInt(name string, defaultValue int) (int, error) {
	val := os.Getenv(name)
	if val == "" {
		return defaultValue, nil
	}
	return strconv.Atoi(val)
}

// AnalyzeTimeoutSeconds returns the number of seconds of how long pg-dba will
// wait for an Analyze operation on the DB.
func AnalyzeTimeoutSeconds() (int, error) {
	return envInt("ANALYZE_TIMEOUT_SECONDS", 600)
}

// DBConnectionString forms & returns the DBConnectionString
func DBConnectionString() string {
	return fmt.Sprintf("host=%s dbname=%s user=%s password=%s sslmode=%s",
		DBHost(),
		DBName(),
		DBUser(),
		DBPassword(),
		SSLMode(),
	)
}

// DBHost returns the POSTGRES_HOST if exists, otherwise localhost
func DBHost() string {
	return envDefault("POSTGRES_HOST", "localhost")
}

// DBName returns the POSTGRES_DB if exists, otherwise postgres
func DBName() string {
	return envDefault("POSTGRES_DB", "postgres")
}

// DBPassword returns the POSTGRES_PASSWORD if exists, otherwise ""
func DBPassword() string {
	return envDefault("POSTGRES_PASSWORD", "\"\"")
}

// DBUser returns the POSTGRES_USER if exists, otherwise postgres
func DBUser() string {
	return envDefault("POSTGRES_USER", "postgres")
}

// FullVacuumTimeoutSeconds returns the number of seconds of how long pg-dba will
// wait for an Full Vacuum operation on the DB.
func FullVacuumTimeoutSeconds() (int, error) {
	return envInt("FULL_VACUUM_TIMEOUT_SECONDS", 600)
}

// LogFormat returns a logrus.Formatter that can be used to configure the
// logger. Defaults to the logrus.TextFormatter.
func LogFormat() logrus.Formatter {
	format := lowerEnvDefault("LOG_FORMAT", "")
	switch format {
	case "json":
		return &logrus.JSONFormatter{}
	}
	return &logrus.TextFormatter{}
}

// LogLevel Returns the logrus log level to use. Defaults to Info
func LogLevel() logrus.Level {
	level := lowerEnvDefault("LOG_LEVEL", "")
	switch level {
	case "debug":
		return logrus.DebugLevel
	case "warn":
		return logrus.WarnLevel
	case "error":
		return logrus.ErrorLevel
	default:
	}
	return logrus.InfoLevel
}

// PostAnalyze returns whether or not to run an analyze after doing a full
// vacuum
func PostAnalyze() bool {
	return envBool("POST_ANALYZE", true)
}

// PreAnalyze returns whether or not to run an analzye to update stats before
// Running a full vacuum
func PreAnalyze() bool {
	return envBool("PRE_ANALYZE", true)
}

// SSLMode returns the SSL_MODE if exists, otherwise the default of require.
// See https://godoc.org/github.com/lib/pq#hdr-Connection_String_Parameters
func SSLMode() string {
	return envDefault("SSL_MODE", "require")
}

// VacuumTimeoutSeconds returns the number of seconds of how long pg-dba will
// wait for a Vacuum operation on the DB.
func VacuumTimeoutSeconds() (int, error) {
	return envInt("VACUUM_TIMEOUT_SECONDS", 600)
}

// Verbose returns whether or not we should run queries in verbose mode.
// Default is false
func Verbose() bool {
	return envBool("VERBOSE", false)
}

// BloatQueryTimeoutSeconds returns the number of seconds of how long
// pg-dba will wait for the bloat query operation on the DB.
func BloatQueryTimeoutSeconds() (int, error) {
	return envInt("BLOAT_QUERY_TIMEOUT_SECONDS", 30)
}
